pkg/mover/chart.go

```diff
@@ -424,8 +430,20 @@ func (cm *ChartMover) pushRewrittenImages(imageChanges []*internal.ImageChange)
 		if change.ShouldPush() {
 			err := retry.Do(
 				func() error {
-					cm.logger.Printf("Pushing %s...\n", change.RewrittenReference.Name())
-					err := cm.targetContainerRegistry.Push(change.Image, change.RewrittenReference)
+					var err error
+					imageToPush := change.RewrittenReference
+
+					// if we know the tag, use that when pushing the image
+					if change.Tag != "" {
+						imageToPush, err = name.NewTag(imageToPush.Context().Name(), name.WithDefaultTag(change.Tag))
+						if err != nil {
+							cm.logger.Printf("Unable to determine the original tag for %s: %w", change.RewrittenReference.Name(), err)
+							imageToPush = change.RewrittenReference
```

Comment:
I wonder how this relates to a similar code I have for the load PR:
~~~go
func refToTag(imageRef name.Reference) (name.Tag, error) {
	// more often that not an name.Reference is actually backed by a name.Tag
	if tag, ok := (imageRef).(name.Tag); ok {
		return tag, nil
	}
	return name.NewTag(fmt.Sprintf("%s:%s", imageRef.Context().Name(), imageRef.Identifier()))
}
~~~
In that case what I have is a reference detected from the chart using the hints. Most of the time the reference is implemented by a Tag, so I can return just the cast.
Otherwise I am building up the full tag name, including the id with `{name}:{id}`, which is not exactly the same you are doing here.

For the *save* step I need no tag translation, the tarball multiwrite call already accepts name references directly.

I am guessing we probably want to be consistent here.

Could you provide an example of the cases handled by this?
I am guessing we are handling having a name tag without digest (`path:tagname`) or with digest (`path:tagname@digest`), or maybe also having a digest without a name (`path:tagname@digest`)?
In each sample case what would be the `Context()..Name()`, `change.Tag` and final `imageToPush Tag`?

Reply:
My thinking is that the possible original image formats:
1. reg.vmware.com/myimage
2. reg.vmware.com/myimage:stable
3. reg.vmware.com/myimage@sha256123abc...

We add an implied `:latest` to the first one. Regardless of if we write them to the chart with the tag or the digest (based on the chart template, if it needs tags, or supports either), we want to push the tagged version.

My code in `pullOriginalImages`:
```
		if change.Digest != originalImage.Identifier() {
			change.Tag = originalImage.Identifier()
		}
```
is about getting the tag from cases 1 and 2 above, but not using the digest in 3.

Your code snippet would do that, too, but it looks like it would turn the digest into a tag of the same name?

`Context().Name()` is always just the `registry/repository`, meaning its the image path *without* the tag or digest.
`change.Tag` should be set, only if we can identify the tag used in the original image reference, either explicitly set, or implicitly with latest. `change.Tag` is empty only if the original image reference has a digest (we don't assumed `latest` tag in that case).
The final `imageToPush` in `pushRewrittenImages` is the rewritten image reference, but using `change.Tag` if that was set. So even if we're writing the digest format in the chart, we're pushing the tagged version (using the tag that was given in the original image reference).

Reply:
For the record, in #103 which was just landed a few minutes ago, the line:
~~~go
return name.NewTag(fmt.Sprintf("%s:%s", imageRef.Context().Name(), imageRef.Identifier()))
~~~
Was removed to we avoid the issue you pointed out when there is no named tag but only the digest on the identifier. The save/load code is about respecting the original tags before we move to rewriting them. And the Tag object is an implementation of the ImageReference, so most of the time, if not always, the image reference will be a Tag. So should be able to rely on the casting code. Then let the rewriting part do its thing, as coded here.

I will revisit if the cast needs to be backed by an alternative if the image reference is not implemented by a tag in some cases.